### PR TITLE
Support `options` param for `requestIdleCallback`

### PR DESCRIPTION
**Pull request by jhen0409**
## Motivation (required)

The `requestIdleCallback` sometimes doesn't work in `Debug JS Remotely` mode if I use real device, the callback will never called. I guess it may be debugger worker and device caused by the time gap, or some cause websocket blocking, so it's just sometimes happening.

I think we can support [options](https://developer.mozilla.org/zh-TW/docs/Web/API/Window/requestIdleCallback#Parameters) for that.

## Test Plan (required)

Added an example `Run requestIdleCallback with timeout option` for Timers of UIExplorer, it use `{ timeout: 100 }` option with burn CPU 100ms, we can see `didTimeout` is true.

### CODE REVIEW

**Comment by jhen0409** [comment]
cc @janicduplessis could you look this? The context is https://github.com/facebook/react-devtools/issues/476, the usage is [here](https://github.com/facebook/react-devtools/blob/master/agent/Bridge.js#L235-L239).

File: Libraries/Core/Timers/JSTimers.js [hunk 1]
```diff
@@ -131,14 +132,24 @@ const JSTimers = {
...
+  requestIdleCallback: function(func : Function, options : ?Object) {
     if (JSTimersExecution.requestIdleCallbacks.length === 0) {
       RCTTiming.setSendIdleEvents(true);
     }
 
     const id = _allocateCallback(func, 'requestIdleCallback');
     JSTimersExecution.requestIdleCallbacks.push(id);
+    if (options && options.timeout) {
+      const currentTime = performanceNow();
+      JSTimers.setTimeout(() => {
```
Comment (janicduplessis) [inline Libraries/Core/Timers/JSTimers.js]: We'll need to cancel this timeout if the idle callback is called before it triggers.

File: Libraries/Core/Timers/JSTimers.js [hunk 2]
```diff
@@ -131,14 +132,24 @@ const JSTimers = {
...
-  requestIdleCallback: function(func : Function) {
+  requestIdleCallback: function(func : Function, options : ?Object) {
     if (JSTimersExecution.requestIdleCallbacks.length === 0) {
       RCTTiming.setSendIdleEvents(true);
     }
 
     const id = _allocateCallback(func, 'requestIdleCallback');
     JSTimersExecution.requestIdleCallbacks.push(id);
+    if (options && options.timeout) {
+      const currentTime = performanceNow();
```
Comment (janicduplessis) [inline Libraries/Core/Timers/JSTimers.js]: I think this time should be taken inside the setTimeout function so it is the current time instead of when the idle callback was created.

**Comment by jhen0409** [comment]
@janicduplessis done, thanks for review. :)

File: Libraries/Core/Timers/JSTimers.js [hunk 3]
```diff
@@ -131,14 +132,42 @@ const JSTimers = {
...
-  requestIdleCallback: function(func : Function) {
+  requestIdleCallback: function(func : Function, options : ?Object) {
     if (JSTimersExecution.requestIdleCallbacks.length === 0) {
       Timing.setSendIdleEvents(true);
     }
 
-    const id = _allocateCallback(func, 'requestIdleCallback');
+    const timeout = options && options.timeout;
+    const id = _allocateCallback(
+      timeout ?
```
Comment (janicduplessis) [inline Libraries/Core/Timers/JSTimers.js]: This should be `timeout != null` so passing timeout: 0 still works.

File: Libraries/Core/Timers/JSTimers.js [hunk 4]
```diff
@@ -131,14 +132,42 @@ const JSTimers = {
...
+        deadline => {
+          const timeoutId = JSTimersExecution.requestIdleCallbackTimeouts[id];
+          if (timeoutId) {
+            JSTimers.clearTimeout(timeout);
```
Comment (janicduplessis) [inline Libraries/Core/Timers/JSTimers.js]: I think this should be `timeoutId` and not `timeout`

File: Libraries/Core/Timers/JSTimers.js [hunk 5]
```diff
@@ -149,6 +178,12 @@ const JSTimers = {
       JSTimersExecution.requestIdleCallbacks.splice(index, 1);
     }
 
+    const timeout = JSTimersExecution.requestIdleCallbackTimeouts[timerID];
```
Comment (janicduplessis) [inline Libraries/Core/Timers/JSTimers.js]: Can you call this `timeoutId` to make it clear it's not the same as the other timeout var we use that is the time of the timeout.

File: Libraries/Core/Timers/JSTimers.js [hunk 6]
```diff
@@ -131,14 +132,42 @@ const JSTimers = {
...
+            JSTimersExecution.requestIdleCallbackTimeouts.splice(id, 1);
+          }
+          return func(deadline);
+        } :
+        func,
+      'requestIdleCallback'
+    );
     JSTimersExecution.requestIdleCallbacks.push(id);
+
+    if (timeout) {
```
Comment (janicduplessis) [inline Libraries/Core/Timers/JSTimers.js]: Same here, `timeout != null`

File: Libraries/Core/Timers/JSTimersExecution.js [hunk 7]
```diff
@@ -45,6 +45,7 @@ const JSTimersExecution = {
   timerIDs: ([] : Array<?number>),
   immediates: [],
   requestIdleCallbacks: [],
+  requestIdleCallbackTimeouts: ([] : Array<?number>),
```
Comment (janicduplessis) [inline Libraries/Core/Timers/JSTimersExecution.js]: I think it would be better to use a Map here with the requestIdleCallback timer id as the key and the timeout timer id as the value.

**Comment by jhen0409** [comment]
@janicduplessis done. :)

**Comment by janicduplessis** [comment]
Looks good, thanks @jhen0409 

@facebook-github-bot shipit

**Comment by jhen0409** [comment]
@janicduplessis, is this ready to ship?

**Comment by janicduplessis** [comment]
Yes, looks like the bot didn't listen let's try again.

**Comment by janicduplessis** [comment]
@facebook-github-bot shipit

**Comment by janicduplessis** [comment]
@ericvicenti Looks like the shipit bot isn't working, could you try landing this?

**Comment by jhen0409** [comment]
@janicduplessis @ericvicenti ping

I just updated the branch.

**Comment by janicduplessis** [comment]
@jhen0409 Sorry about the delay, importing seems to fail a lot recently. Let me try again...

**Comment by janicduplessis** [comment]
@facebook-github-bot shipit

**Comment by jhen0409** [comment]
@janicduplessis could you try again? Thanks! 😂

**Comment by janicduplessis** [comment]
@hramos Could you check why this isn't landing?

**Comment by jhen0409** [comment]
Fixed conflict: moved the timer example to RNTester

**Comment by jhen0409** [comment]
@ericnakagawa could you check this? Thanks!

**Comment by jhen0409** [comment]
@hramos could you check this again? Thank you!

I've updated branch and checked it works.

**Comment by gaearon** [comment]
Is anything holding this from landing?

**Comment by hramos** [comment]
@gaearon the import tool initially had some issues importing the PR, but it looks like my last attempt  7 days ago worked (and I missed this). The internal diff should be landing soon.

**Comment by gaearon** [comment]
Thanks!
